Add extra value validation for matchExpression field in LabelSelector

liggitt opened the pull request:
#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:

fixes https://github.com/kubernetes/kubernetes/issues/107415

fixes https://github.com/kubernetes/kubernetes/issues/99139

#### Special notes for your reviewer:

Almost all of the fixes were made by @Zheaoli in https://github.com/kubernetes/kubernetes/pull/107531 

This PR adds a commit addressing comments needed for merge in order to meet code freeze deadlines. New APIs are starting to use the label selector validation, and we don't want to expand the problem in 1.26

#### Does this PR introduce a user-facing change?
```release-note
metav1.LabelSelectors specified in API objects are now validated to ensure they do not contain invalid label values that will error at time of use. Existing invalid objects can be updated, but new objects are required to contain valid label selectors.
```

/sig api-machinery
/cc @jpbetz

Zheaoli committed:
Add extra value validation for matchExpression field in LabelSelector

liggitt committed:
Update workload selector validation

jpbetz reviewed (commented):
I did an independent audit looking for any types that might have gotten missed. I couldn't find any (HorizontalPodAutoscalerSpec does not have Metrics in v1). 

As we already discussed offline, this will need to handle the selectors in https://github.com/kubernetes/kubernetes/pull/113314 once it merges.

liggitt commented on staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/validation/validation.go:
```diff
@@ -28,19 +28,42 @@ import (
+	}
+	return false
+}
+
+func ValidateLabelSelector(ps *metav1.LabelSelector, opts LabelSelectorValidationOptions, fldPath *field.Path) field.ErrorList {
```
documented and clarified that the "allowInvalid" option is only expected to be used when required for backwards compatibility

liggitt commented:
resolved unrelated rebase conflict in pkg/registry/batch/job/strategy_test.go, updated godoc comments and reverted the test change in https://github.com/kubernetes/kubernetes/pull/113699#discussion_r1015979766

liggitt commented:
/hold for in-flight PR adding another new callsite... will rebase once https://github.com/kubernetes/kubernetes/pull/113314 merges and update those calls to unconditionally use the strict validation, since they are new in 1.26

liggitt commented:
rebased after https://github.com/kubernetes/kubernetes/pull/113314 merged

/hold cancel

liggitt commented:
/skip

jpbetz commented:
/lgtm
Thanks for fixing this for CEL admission control! It feels good to know all the fields are fully validated before the initial release of the new API!

liggitt commented:
/retest

pacoxu commented:
/retest
